This patch is generated from the squid3-icap branch of HEAD in squid3
Fri Jun 29 00:16:28 2007 GMT
See http://devel.squid-cache.org/

Index: squid3/src/http.cc
diff -u squid3/src/http.cc:1.115 squid3/src/http.cc:1.49.2.78
--- squid3/src/http.cc:1.115	Mon Jun 25 17:51:06 2007
+++ squid3/src/http.cc	Wed Jun 27 19:54:46 2007
@@ -1904,8 +1904,19 @@
 HttpStateData::handleRequestBodyProducerAborted()
 {
     ServerStateData::handleRequestBodyProducerAborted();
-    // XXX: SendComplete(COMM_ERR_CLOSING) does little. Is it enough?
     SendComplete(fd, NULL, 0, COMM_ERR_CLOSING, 0, this);
+    if (entry->isEmpty()) {
+        debugs(11,9, HERE << "creating error entry for producer abort");
+        // XXX: This may not be an ICAP error. The error is probably
+        // irrelevant if it was the HTTP client that aborted the request.
+        // We use ERR_ICAP in hope that all other cases are ICAP-related.
+        // FwdState asserts there is an error.
+        ErrorState *err =
+            errorCon(ERR_ICAP_FAILURE, HTTP_INTERNAL_SERVER_ERROR, orig_request);
+        fwd->fail(err);
+    }
+
+    abortTransaction("premature end of request body while sending");
 }
 
 // called when we wrote request headers(!) or a part of the body
